### DISCUSSION TITLE
Kanidm: What does `endpoint belongs to different authority ` actually mean?

### DISCUSSION THREAD

**Discussion by TheRealGramdalf** [9p4/jellyfin-plugin-sso#156] [Q&A] [answered]
When attempting to use Kanidm as an identity provider for jellyfin (via OIDC), it throws an `endpoint belongs to different authority ` error. Though mentioned several times (see: Readme, https://github.com/9p4/jellyfin-plugin-sso/issues/153, https://github.com/9p4/jellyfin-plugin-sso/issues/113, https://github.com/9p4/jellyfin-plugin-sso/issues/136), it isn't quite clear to me what this actually means. 

I believe it relates to the API path on the IdP used by the resource server (jellyfin). The Kanidm documentation says:
>Kanidm will expose its OAuth2 APIs at the following URLs:
    - user auth url: https://idm.example.com/ui/oauth2
    - api auth url: https://idm.example.com/oauth2/authorise
    - token url: https://idm.example.com/oauth2/token
    - rfc7662 token introspection url: https://idm.example.com/oauth2/token/introspect
    - rfc7009 token revoke url: https://idm.example.com/oauth2/token/revoke

For reference, here's my jellyfin/kanidm configuration:
![image](https://github.com/9p4/jellyfin-plugin-sso/assets/79593869/3ddf7356-941a-4c6b-89a2-20ae51214821)

And the `.well-known`:
```json
{
  "issuer": "https://auth.domain.tld/oauth2/openid/jellyfin-aer",
  "authorization_endpoint": "https://auth.domain.tld/ui/oauth2",
  "token_endpoint": "https://auth.domain.tld/oauth2/token",
  "userinfo_endpoint": "https://auth.domain.tld/oauth2/openid/jellyfin-aer/userinfo",
  "jwks_uri": "https://auth.domain.tld/oauth2/openid/jellyfin-aer/public_key.jwk",
  "scopes_supported": [
    "openid",
    "profile"
  ],
  "response_types_supported": [
    "code"
  ],
  "response_modes_supported": [
    "query"
  ],
  "grant_types_supported": [
    "authorization_code"
  ],
  "subject_types_supported": [
    "public"
  ],
  "id_token_signing_alg_values_supported": [
    "ES256"
  ],
  "token_endpoint_auth_methods_supported": [
    "client_secret_basic",
    "client_secret_post"
  ],
  "display_values_supported": [
    "page"
  ],
  "claim_types_supported": [
    "normal"
  ],
  "service_documentation": "https://kanidm.github.io/kanidm/master/integrations/oauth2.html",
  "claims_parameter_supported": false,
  "request_parameter_supported": true,
  "request_uri_parameter_supported": false,
  "require_request_uri_registration": false
}
```

**Comment by 9p4** [comment]
Try pressing the checkbox "Do Not Validate OpenID Endpoints (Insecure)"

**Reply by 9p4** [reply]
This usually means something is misconfigured on the reverse proxy, but it can vary and depends on your setup.

**Reply by TheRealGramdalf** [reply]
Checking the box does indeed work, but I would still like to know what it actually means if possible. 
I'm using traefik as an RP, which terminates SSL/TLS connection with a wildcard cert (`*.domain.tld`) and connects to Kanidm over SSL/TLS, using Kanidm's self-signed cert. From what I know it could be an SSL/TLS issue, due to multiple certs, or it could be how jellyfin is accessing Kanidm - I have a DNS override that points to the internal IP of traefik, which then proxies the connection to Kanidm. The setup is quite complicated and I haven't quite nailed down some details, so at the moment it's still in flux.

**Comment by ToxicMushroom** [comment]
Ran into the same issue, but my kanidm instance is not behind traefik TLS stuff, so it probably is unrelated from traefik ? (I do still use it to route tcp with tls pass-trough)

_Kanidm does work with mastodon but I also wouldn't call their current implementation a perfect example._

**Reply by 9p4** [reply]
Take a look at your `.well-known/openid-configuration` path on Kanidm, and ensure that the URLs returned match the domain that Kanidm is running on.

**Reply by ToxicMushroom** [reply]
they do https://idm.melijn.com/oauth2/openid/jellyfin/.well-known/openid-configuration

**Reply by 9p4** [reply]
Does your configuration reference `https://idm.melijn.com/oauth2/openid/jellyfin` in the configuration?

**Reply by Firstyear** [reply]
It appears that the check is coming from https://github.com/IdentityModel/IdentityModel/blob/0d5186eb846df86b1228fdfb73dceed7d69d9ad0/src/Client/Messages/DiscoveryDocumentResponse.cs#L200 

I wonder if the issue here is that .Authority on uri returns dns name *and* port, per https://learn.microsoft.com/en-us/dotnet/api/system.uri.authority?view=net-8.0 

Maybe IdentityModel needs to say something like `Endpoint is on a different host than authority: {endpoint} expected {host}`. That way it's easier to see what part of the config is incorrect, but also what is expected?

**Reply by 9p4** [reply]
It is because the endpoint is `https://idm.melijn.com/oauth2/openid/jellyfin` but the well-known returns URLs that do not start with that endpoint path. This is a (very minor) design issue on ~~Kanidm's~~IdentityModel's part.

**Reply by Firstyear** [reply]
https://openid.net/specs/openid-connect-discovery-1_0.html does not require relative locations of urls. In fact it seems the only requirement is that `iss` is the prefix to `.well-known` (section 4.3), which we do correctly. 

So I don't really see how this is a design issue in Kanidm, but more that an incorrect assumption is being made in IdentityModel.

**Reply by 9p4** [reply]
For additional reference: https://github.com/IdentityModel/IdentityModel/issues/456

**Reply by Firstyear** [reply]
That's a reasonable check for the authority, but it's not checking only the authority, but and entire section of the uri matches the issuer parameter. So perhaps a new issue needs to be raised?

**Reply by 9p4** [reply]
Do you think that the documentation for the plugin ought to recommend turning off the endpoint validation? There is a checkbox for it in the plugin configuration that is on by default.

**Reply by Firstyear** [reply, marked as answer]
The problem is that the validation is validating "too much". It's validation to enforce that it's the same URI origin is good, but the fact that it expects all the URI to be relative to issuer is the problem here. But if you disable that check (to allow non-relative to issuer URI) you also disable the URI origin verification. 

So you could change that in documentation, but I would advise and prefer to see that the validation only check the URI origin is consistent, and not that URI's are relative to issuer. Else a potentially useful security feature is then off by default.

**Reply by 9p4** [reply]
I think this may be something that has to be taken up with IdentityModel

**Reply by Firstyear** [reply]
I agree, would you like to open the issue with the context here since you are the maintainer of this project?

**Reply by 9p4** [reply]
I believe you would be better qualified to describe the issue and be a liaison with their developers, since you are the one who maintains Kanidm.

**Reply by Firstyear** [reply]
https://github.com/IdentityModel/IdentityModel/issues/553
